### DISCUSSION TITLE
BNS library feature scope

### DISCUSSION THREAD

**Discussion by agraebe** [stx-labs/stacks.js#1038] [General]
Right now, the BNS library includes methods to interact with the [BNS contract](https://docs.stacks.co/references/bns-contract). With the recent addition of [BNS API endpoints](https://blockstack.github.io/stacks-blockchain-api/#tag/bns), there are several new fetching capabilities that are not directly implemented in the BNS contract (and not in the BNS library).

**The question is: Should we extend the scope of the BNS library and include new capabilities provided by the API?**

I would argue that developers using BNS shouldn't need to worry about the difference between the contract, API, or library. This is an additional burden to the onboarding flow that we can avoid to improve the developer experience. It would be great if the BNS library could be the go-to for all BNS features developers would care about.

In terms of BNS features that developers care about today, there's some signal from current research:
- Better UX: would like to display a human-readable username instead of the stacks address
- Simplified interactions: would like to send tokens to/from usernames instead of the stacks address

Another thing to note is that Hiro wallets are also not allowing users to register a username during the account creation process. The only registrar currently available is btc.us.

All things considered, I would assume that developers would mainly care about fetching capabilities (eg get username for an account). This would reason them _not_ to use the BNS library and integrate the BNS API endpoints instead. Thus, limiting the impact of the BNS library. If we believe this is OK, and the scope should stay as is, we should explicitly communicate the scope of the library. For instance, BNS JS library for ...
- contract interaction
- writing operations
- registrar features

**Comment by kyranjamie** [comment]
Related https://github.com/blockstack/stacks.js/issues/994

tl;dr: no

I opened this a while back as, at least in the use case described in this issue, we came across some difficulties owing to API requests being handled through stacks.js libs. We may well come across similar issues in future. 

If we do allow users to invoke API requests through stacks.js libs, **we must let the consumer configure all properties of the request**. So this configuration-passing functionality also needs to be built into the BNS library.

Further, it also means that the library determines what http client lib the consumer uses. What if the package consumer want to use a different one? Do they have to bundle the one chosen here anyway? We don't know the environment this is being used, so you're forced to use `cross-fetch` and libraries like this. Consumer, _can_ change the lib if it conforms to `fetch`s API, but they must to do so with their package-bundling tooling. A PITA.

My suggestion would be to prohibit all transport layer interaction in the BNS library (and most other packages), no http requests or side-effecty behaviour. Instead we could expose methods to help _prepare_ these API requests, but not actually call them within. Can `@stacks/bns` play nice with our generated API client?

**Reply by agraebe** [reply]
great points - maybe worth untangling a few things:

1. Should the bns library make it easy to interact with the features bns provides (independent of _how_ it would do that)? It seems we would prefer that
2. Should stacks.js packages _in general_ enable devs to configure the request library and options? It seems we would prefer that as well. The API client library uses a pattern to define the fetching library, so maybe something similar to that would be a path forward?
3. Should stacks.js packages (eg stacking, bns) _in general_ make use of the generated API client? I think we wanted to explore and discuss this further. I believe we didn't find a solution yet

Does it seem a good step could be to hash out how the generated API client could potentially be used inside of stacks.js packages? This library might enable the configuration options we're seeking and reduce code duplication. Thoughts?

**Reply by kyranjamie** [reply]
To answer your questions:
1) Yes. Though encapsulating all functionality in a single package isn't the _only_ way to do this.
2) If making network requests, then yes, its necessary to allow consumers to configure request. That's possible, but also quite some work to implement this in all packages that handle requests, I wouldn't recommend that path.
3) No. I don't think they should. 

I didn't explain very clearly, my suggestion is **not to put API requests in core packages at all**.
#994 highlights issues caused by coupling API request logic in our core package utilities. Adding in HTTP request/client configuration functionality is more of a workaround than a solution.

Here's a fun ASCII diagram that illustrates the package dependency set up I'd advocate:

```
   +--------------------------+                                                
   |                          |                                                
   |                          |-\                                              
   |       @stacks/bns        |  ---\                                          
   |                          |      --\                                       
   |                          |         ---\          +-----------------------+
   +--------------------------+             --\       |                       |
                                               ---\   |   @stacks/bns-sdk     |
   +--------------------------+                   ----|                       |
   |                          |            ------/    |  developer-script.js  |
   |                          |    -------/           |                       |
   |  @stacks/{api/network}   |---/                   +-----------------------+
   |                          |                                                
   |                          |                                                
   +--------------------------+                                                
```

We should compose the packages in a way where core packages, like `bns` or `transaction` are concerned with _preparing_ transactions, but not cause side-effects, like broadcasting them

## Option 1 — A "smarter" SDK package
```ts
import { BnsClient } from '@stacks/bns-sdk';
const bns = new BnsClient().allTheMethodsYouNeedForBnsIncludingApiRequests();
```

Easy to use, but not necessary imo. 

## Option 2 — Developers compose packages themselves
```ts
import { buildPreorderNameTx } from '@stacks/bns';
import { broadcastTransaction } from '@stacks/api'; // or generated client lib

const preorderNameTx = await buildPreorderNameTx({ ...args });
const resp = await broadcastTransaction(preorderNameTx);
```

I prefer this option:
- Dev can compose what they need. We don't make assumptions on what HTTP client they use, how they configure their reqs etc
- `@stacks/bns` is a library concerned only with contract interaction fns with no side-effects

Hope I don't get accused of bike-shedding here lol, I just think we'd benefit from each package having slightly more scoped responsibilities.

**Reply by aulneau** [reply]
> Hope I don't get accused of bike-shedding here lol, I just think we'd benefit from each package having slightly more scoped responsibilities.

This is a fantastic write up and in no way is it bikeshedding. These are all important things to consider and a direction I think we should go in.

**Reply by agraebe** [reply]
Agree with @aulneau! Not bikeshedding at all. It's a fantastic suggestion and I personally would also prefer option 2 after giving it more thought. @reedrosenbluth - any thoughts from your end?

**Reply by reedrosenbluth** [reply]
Awesome writeup @kyranjamie! I'm definitely in favor of this.

**Reply by zone117x** [reply]
> My suggestion would be to prohibit all transport layer interaction in the BNS library (and most other packages), no http requests or side-effecty behaviour. Instead we could expose methods to help prepare these API requests, but not actually call them within

I agree, more explicit and less side effects sounds better. I think something like option 2 makes sense, but would be useful to have a more full example, because it might look like a lot more steps. For example, a network request needs to be made to get the fee, then to get the nonce, then build a final transaction, then a network request to broadcast the tx.

I'd predict complaints from devs if the "all-in-one" functions were removed, and a single function call had to be replaced by like 5 lines of code.

**Comment by friedger** [comment]
Using the node api client library inside stacks.js means that developers can only switch to a different node url if that node runs the stacks api server. This might not be the case. Therefore, I would try to create client libraries with clear requirements on the server. Probably developers run a custom stacks api server specific to their contracts.

**Comment by aulneau** [comment]
I would agree with @kyranjamie and try to have very clear definitions of what each package does. I think we should generally move away of having "one package to rule them all" for a given category of functionality. The most complex and confusing things about BNS are how to construct the arguments needed for the various contract functions.

> Better UX: would like to display a human-readable username instead of the stacks address
> Simplified interactions: would like to send tokens to/from usernames instead of the stacks address

To me, these don't seem related to the BNS library at all, but instead these should be exposed via connect and the transactions lib. Perhaps these libraries would implement some BNS-sdk package.
